runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt

```diff
@@ -0,0 +1,182 @@
+package kotlinx.serialization
+
+import kotlinx.serialization.internal.StringDescriptor
+import kotlinx.serialization.internal.nullable
+import kotlinx.serialization.json.internal.BEGIN_LIST
+import kotlinx.serialization.json.internal.BEGIN_OBJ
+import kotlinx.serialization.json.internal.END_LIST
+import kotlinx.serialization.json.internal.END_OBJ
+import kotlinx.serialization.modules.EmptyModule
+import kotlinx.serialization.modules.SerialModule
+import kotlinx.serialization.modules.getContextualOrDefault
+
+
+/**
+ * Converts Kotlin data structures to plain Javascript objects
+ *
+ *
+ * Limitations:
+ * * Map keys must be of primitive or enum type
+ * * Custom serializers are not supported
+ * * Enums are serialized as the value of `@SerialName` if present or their name, in that order. Ordinals are not supported.
+ * * There is no analogue to DynamicObjectParsers' useArrayPolymorphism flag
+ *
+ * @param encodeDefaults if true any property which equals exactly its default value will be omitted
+ * @param encodeNullAsUndefined if true null properties will be omitted from the output
+ */
+class DynamicObjectSerializer(
+    private val encodeDefaults: Boolean = true,
+    private val encodeNullAsUndefined: Boolean = false,
```

Comment:
Maybe create a configuration class, like `JsonConfiguration`.

Reply:
It seems like an overkill to me. I'd rather discuss whether we need the options in the first place. It's hard for me to imagine a use case for `encodeDefaults` (in the context of DynamicObjectSerializer) and while I have a use case for `encodeNullAsUndefined` it is also a bit out there.

My use-case for `encodeNullAsUndefined`: I want to pass some data to a table component (e.g. https://material-table.com/#/). I know (that for some business reason) some property will always be null, therefore I use `encodeNullAsUndefined ` to omit nulls which means there will be no column for that property in the table (instead of an empty one).

Reply:
I think it is fine to have two separate flags. Maybe it's even ok to put full `JsonConfiguration` itself here, see https://github.com/Kotlin/kotlinx.serialization/pull/694/files#r376207786 , but I do not insist.